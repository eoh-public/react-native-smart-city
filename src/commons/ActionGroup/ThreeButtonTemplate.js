import React from 'react';
import { View, StyleSheet, TouchableOpacity } from 'react-native';
import { Icon } from '@ant-design/react-native';

import Text from '../Text';
import { Colors } from '../../configs';

const ThreeButtonTemplate = ({ actionGroup, doAction }) => {
  const { configuration } = actionGroup;

  return (
    <View style={styles.wrap}>
      <TouchableOpacity
        style={styles.buttonActionDoor}
        onPress={() => doAction(configuration.action1_data)}
        underlayColor="#FAFAFA"
      >
        <View style={styles.imageBtn}>
          <Icon name={configuration.icon1} size={30} color={Colors.Gray5} />
        </View>
        <Text style={styles.text}>{configuration.text1}</Text>
      </TouchableOpacity>

      <TouchableOpacity
        style={styles.buttonActionDoor}
        onPress={() => doAction(configuration.action2_data)}
        underlayColor="#FAFAFA"
      >
        <View style={styles.imageBtn}>
          <Icon name={configuration.icon2} size={30} color={Colors.Gray5} />
        </View>
        <Text style={styles.text}>{configuration.text2}</Text>
      </TouchableOpacity>

      <TouchableOpacity
        style={styles.buttonActionDoor}
        onPress={() => doAction(configuration.action3_data)}
        underlayColor="#FAFAFA"
      >
        <View style={styles.imageBtn}>
          <Icon name={configuration.icon3} size={30} color={Colors.Gray5} />
        </View>
        <Text style={styles.text}>{configuration.text3}</Text>
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  wrap: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    paddingHorizontal: 8,
    paddingVertical: 16,
    marginHorizontal: 20,
    marginTop: 20,
    backgroundColor: Colors.White,
    borderWidth: 1,
    borderColor: Colors.Gray4,
    borderRadius: 10,
    shadowColor: Colors.Black,
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 3,
  },
  imageBtn: {
    height: 28,
    marginBottom: 7,
  },
  text: {
    fontSize: 14,
    lineHeight: 22,
  },
  buttonActionDoor: {
    flex: 1,
    aspectRatio: 1,
    marginHorizontal: 4,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: Colors.White,
    borderWidth: 1,
    borderColor: Colors.Gray4,
    borderRadius: 10,
    shadowColor: Colors.Shadow,
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 3,
  },
});

export default ThreeButtonTemplate;
