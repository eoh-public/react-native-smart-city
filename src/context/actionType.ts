export const Action = {
  UPDATE_AUTH: 'UPDATE_AUTH',
  UPDATE_LANGUAGE: 'UPDATE_LANGUAGE',
  STORE_STATUS_BAR: 'STORE_STATUS_BAR',
  SET_CONFIG: 'SET_CONFIG',
};

export type AuthData = {
  account: {
    token: string;
    user: any;
  };
};

export type ConfigData = {
  LG_CLIENT_ID: string;
  LG_REDIRECT_URI_APP: string;
};

export type Language = 'en' | 'vi';

export type StatusBar = {
  backgroundColor: string;
  barStyle: string;
};

export type ActionType = keyof typeof Action;

export type ActionDataMap = {
  UPDATE_AUTH: AuthData;
  UPDATE_LANGUAGE: Language;
  STORE_STATUS_BAR: StatusBar;
  SET_CONFIG: ConfigData;
};
