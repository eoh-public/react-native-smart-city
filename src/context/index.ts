export { useSCContextSelector, SCContext, SCProvider } from './SCContext';
