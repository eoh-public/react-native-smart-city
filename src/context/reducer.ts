import { initData } from '../utils/InitData';
import { ActionType, AuthData, Language, StatusBar } from './actionType';

export type ContextData = {
  auth: AuthData;
  language: Language;
};

export type Action = {
  type: ActionType;
  payload: any;
};

export const initialState = {
  auth: {
    account: {
      token: '',
      user: {},
    },
  },
  language: 'en' as Language,
  statusBar: {} as StatusBar,
};

export const reducer = (currentState: ContextData, action: Action) => {
  const { type, payload } = action;
  switch (type) {
    case 'UPDATE_AUTH':
      initData((payload as AuthData).account);
      return { ...currentState, auth: payload };
    case 'STORE_STATUS_BAR':
      return { ...currentState, statusBar: payload };
    default:
      return currentState;
  }
};
