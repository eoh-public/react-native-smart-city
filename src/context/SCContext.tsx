import React, { useContext, useReducer } from 'react';
import { StyleSheet, View, Text } from 'react-native';
import Toast from 'react-native-toast-message';

import { Alert } from '../commons';
import { Colors } from '../configs';
import {
  ActionDataMap,
  ActionType,
  AuthData,
  Language,
  StatusBar,
} from './actionType';
import { initialState, Action, ContextData, reducer } from './reducer';

const toastConfig = {
  // only for error for now
  error: (internalState) => (
    <View style={styles.toastContainer}>
      <Text style={styles.textWhite}>{internalState.text1}</Text>
    </View>
  ),
};

type SCContextType = {
  stateData: ContextData;
  setAuth: (authData: AuthData) => void;
  setLocale: (language: Language) => void;
  setStatusBar: (statusBar: StatusBar) => void;
  updateLocale: (langTranslate: any) => void;
  setAction: <T extends ActionType>(
    action: T,
    payload?: ActionDataMap[T]
  ) => void;
};

export const SCContext = React.createContext<SCContextType>(
  {} as SCContextType
);
type Reducer<StateData, Action> = (
  state: StateData,
  action: Action
) => StateData;

export const SCProvider = ({ children }) => {
  const [stateData, dispatch] = useReducer<Reducer<ContextData, Action>>(
    reducer,
    initialState
  );

  const setAuth = (authData: AuthData) => {
    setAction('UPDATE_AUTH', authData);
  };

  const setStatusBar = (statusBar: StatusBar) => {
    setAction('STORE_STATUS_BAR', statusBar);
  };

  const setLocale = (language: Language) => {
    setAction('UPDATE_LANGUAGE', language);
  };

  const setAction = <T extends ActionType>(
    action: T,
    payload?: ActionDataMap[T]
  ) => {
    dispatch({ type: action, payload: payload });
  };

  const providerValue = {
    stateData,
    setAuth,
    setAction,
    setLocale,
    setStatusBar,
  };

  return (
    <SCContext.Provider value={providerValue}>
      {children}
      <Alert ref={(ref) => Alert.setRef(ref)} />
      <Toast config={toastConfig} ref={(ref) => Toast.setRef(ref)} />
    </SCContext.Provider>
  );
};

export const useSCContextSelector = (
  selector: (contextData: ContextData) => unknown
) => {
  const { stateData } = useContext(SCContext);
  return selector(stateData);
};

const styles = StyleSheet.create({
  toastContainer: {
    borderRadius: 5,
    padding: 10,
    backgroundColor: Colors.Black,
  },
  textWhite: {
    color: Colors.White,
  },
});
