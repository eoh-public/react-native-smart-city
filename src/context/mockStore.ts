import { ContextData } from './reducer';

export const mockDataStore: ContextData = {
  auth: {
    account: {
      token: 'TOKEN',
      user: {
        avatar: null,
        email: 'test@gmail.com',
        id: 1,
        is_using_social_avatar: false,
        name: 'TEST',
        org: 2,
        phone_number: '012345678',
        social_avatar_hash: '',
      },
    },
  },
  language: 'en',
  listDevice: {
    sentEmail: false,
  },
  statusBar: {
    backgroundColor: '',
    barStyle: '',
  },
  listAction: [],
};

export const mockSCStore = (data: ContextData): ContextData => {
  return {
    auth: {
      account: {
        ...mockDataStore.auth.account,
        ...data.auth?.account,
      },
    },
    language: data.language ? data.language : mockDataStore.language,
    listDevice: {
      ...mockDataStore.listDevice,
      ...data.listDevice,
    },
    statusBar: {
      backgroundColor:
        data?.statusBar?.backgroundColor ||
        mockDataStore?.statusBar?.backgroundColor,
      barStyle: data?.statusBar?.barStyle || mockDataStore?.statusBar?.barStyle,
    },
    listAction: [...mockDataStore.listAction, ...(data?.listAction || [])],
  };
};
