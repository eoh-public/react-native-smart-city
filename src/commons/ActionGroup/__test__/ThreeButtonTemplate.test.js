import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { TouchableOpacity } from 'react-native';

import ThreeButtonTemplate from '../ThreeButtonTemplate';
import { SCProvider } from '../../../context';
import { mockSCStore } from '../../../context/mockStore';

const wrapComponent = (actionGroup) => (
  <SCProvider initState={mockSCStore({})}>
    <ThreeButtonTemplate actionGroup={actionGroup} doAction={jest.fn()} />
  </SCProvider>
);

describe('Test ThreeButtonTemplate', () => {
  const actionGroup = {
    configuration: {
      action1: '2b949045-8e03-4c07-a855-7794ade2e69c',
      action1_data: {
        color: '#00979D',
        command_prefer_over_bluetooth: true,
        command_prefer_over_googlehome: false,
        command_prefer_over_internet: false,
        googlehome_actions: [],
        icon: 'caret-up',
        id: 9,
        key: '2b949045-8e03-4c07-a855-7794ade2e69c',
      },
      action2: '38347d5e-4418-4ab0-978c-c82f4c034897',
      action2_data: {
        color: '#00979D',
        command_prefer_over_bluetooth: true,
        command_prefer_over_googlehome: false,
        command_prefer_over_internet: false,
        googlehome_actions: [],
        icon: 'stop',
        id: 11,
        key: '38347d5e-4418-4ab0-978c-c82f4c034897',
      },
      action3: 'a492e08c-8cb1-44ee-8ea0-46aaca4e5141',
      action3_data: {
        color: '#00979D',
        command_prefer_over_bluetooth: true,
        command_prefer_over_googlehome: false,
        command_prefer_over_internet: false,
        googlehome_actions: [],
        icon: 'caret-down',
        id: 10,
        key: 'a492e08c-8cb1-44ee-8ea0-46aaca4e5141',
      },
      icon1: 'caret-up',
      icon2: 'stop',
      icon3: 'caret-down',
      text1: 'UP',
      text2: 'STOP/UNLOCK',
      text3: 'DOWN',
      id: 1,
      template: 'three_button_action_template',
    },
  };

  let wrapper;

  test('render ThreeButtonTemplate', async () => {
    await act(() => {
      wrapper = renderer.create(wrapComponent(actionGroup));
    });
    expect(wrapper.toJSON()).toMatchSnapshot();
    const instance = wrapper.root;
    const touchableOpacities = instance.findAllByType(TouchableOpacity);
    expect(touchableOpacities.length).toEqual(3);
  });
});
