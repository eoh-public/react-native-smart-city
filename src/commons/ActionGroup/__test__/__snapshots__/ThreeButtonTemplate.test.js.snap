// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ThreeButtonTemplate render ThreeButtonTemplate 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "borderColor": "#E8E8E8",
      "borderRadius": 10,
      "borderWidth": 1,
      "elevation": 3,
      "flexDirection": "row",
      "justifyContent": "space-around",
      "marginHorizontal": 20,
      "marginTop": 20,
      "paddingHorizontal": 8,
      "paddingVertical": 16,
      "shadowColor": "#000000",
      "shadowOffset": Object {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.25,
      "shadowRadius": 3.84,
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "aspectRatio": 1,
        "backgroundColor": "#FFFFFF",
        "borderColor": "#E8E8E8",
        "borderRadius": 10,
        "borderWidth": 1,
        "elevation": 3,
        "flex": 1,
        "justifyContent": "center",
        "marginHorizontal": 4,
        "opacity": 1,
        "shadowColor": "#303133",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
      }
    }
  >
    <View
      style={
        Object {
          "height": 28,
          "marginBottom": 7,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#D9D9D9",
              "fontFamily": "antoutline",
              "fontSize": 30,
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            undefined,
          ]
        }
      >
        
      </Text>
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontFamily": "SFProDisplay-Regular",
            "fontSize": undefined,
            "lineHeight": undefined,
            "textAlign": "left",
            "textDecorationLine": "none",
          },
          Object {
            "fontSize": 14,
            "lineHeight": 22,
          },
        ]
      }
    >
      UP
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "aspectRatio": 1,
        "backgroundColor": "#FFFFFF",
        "borderColor": "#E8E8E8",
        "borderRadius": 10,
        "borderWidth": 1,
        "elevation": 3,
        "flex": 1,
        "justifyContent": "center",
        "marginHorizontal": 4,
        "opacity": 1,
        "shadowColor": "#303133",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
      }
    }
  >
    <View
      style={
        Object {
          "height": 28,
          "marginBottom": 7,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#D9D9D9",
              "fontFamily": "antoutline",
              "fontSize": 30,
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            undefined,
          ]
        }
      >
        
      </Text>
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontFamily": "SFProDisplay-Regular",
            "fontSize": undefined,
            "lineHeight": undefined,
            "textAlign": "left",
            "textDecorationLine": "none",
          },
          Object {
            "fontSize": 14,
            "lineHeight": 22,
          },
        ]
      }
    >
      STOP/UNLOCK
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "aspectRatio": 1,
        "backgroundColor": "#FFFFFF",
        "borderColor": "#E8E8E8",
        "borderRadius": 10,
        "borderWidth": 1,
        "elevation": 3,
        "flex": 1,
        "justifyContent": "center",
        "marginHorizontal": 4,
        "opacity": 1,
        "shadowColor": "#303133",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
      }
    }
  >
    <View
      style={
        Object {
          "height": 28,
          "marginBottom": 7,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#D9D9D9",
              "fontFamily": "antoutline",
              "fontSize": 30,
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            undefined,
          ]
        }
      >
        
      </Text>
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontFamily": "SFProDisplay-Regular",
            "fontSize": undefined,
            "lineHeight": undefined,
            "textAlign": "left",
            "textDecorationLine": "none",
          },
          Object {
            "fontSize": 14,
            "lineHeight": 22,
          },
        ]
      }
    >
      DOWN
    </Text>
  </View>
</View>
`;
